### DISCUSSION TITLE
IKOS execution problems on the code base

### DISCUSSION THREAD

**Discussion by ThibFrgsGmz** [nasa/fprime#1792] [Q&A]
Hi everyone,

Has anyone tried or successfully run [NASA V&V's IKOS](https://github.com/NASA-SW-VnV/ikos) on the F´ code base?

I'm trying to run it to see its suitability and propose its integration into a dedicated CI workflow.
Since it is [planned to use](https://github.com/ament/ament_ikos) IKOS on [SpaceROS](https://github.com/space-ros), I suspect that it is relevant to deploy it. See https://github.com/ament/ament_lint/pull/325

Personally, I was unlucky in my experience, I had an error when generating the CMake cache, see below.

Before contacting the IKOS team I wanted to know if you had done the experiment.
The IKOS team doesn't seem to be very active lately (the latest versions of LLVM are not supported).

To scan a code base, we need the [ikos-scan](https://github.com/NASA-SW-VnV/ikos/blob/master/analyzer/README.md#analyze-a-whole-project-with-ikos-scan) utility which overrides the `CC` and `CXX` environment variables to intercept the compiler commands.

```
$ ikos-scan fprime-util generate
[INFO] Generating build directory at: /home/ThibFrgsGmz/fprime/Ref/build-fprime-automatic-native
[INFO] Using toolchain file None for platform default
-- The C compiler identification is Clang 9.0.1
-- The CXX compiler identification is Clang 9.0.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /home/ThibFrgsGmz/ikos/build/install/bin/ikos-scan-cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - failed
-- Check for working CXX compiler: /home/ThibFrgsGmz/ikos/build/install/bin/ikos-scan-c++
-- Check for working CXX compiler: /home/ThibFrgsGmz/ikos/build/install/bin/ikos-scan-c++ - broken
CMake Error at /usr/share/cmake-3.25/Modules/CMakeTestCXXCompiler.cmake:63 (message):
-- Configuring incomplete, errors occurred!
  The C++ compiler
See also "/home/ThibFrgsGmz/fprime/Ref/build-fprime-automatic-native/CMakeFiles/CMakeOutput.log".

See also "/home/ThibFrgsGmz/fprime/Ref/build-fprime-automatic-native/CMakeFiles/CMakeError.log".
    "/home/ThibFrgsGmz/ikos/build/install/bin/ikos-scan-c++"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/ThibFrgsGmz/fprime/Ref/build-fprime-automatic-native/CMakeFiles/CMakeScratch/TryCompile-S9pOWR

    Run Build Command(s):/usr/bin/make -f Makefile cmTC_e755a/fast && /usr/bin/make  -f CMakeFiles/cmTC_e755a.dir/build.make CMakeFiles/cmTC_e755a.dir/build
    make[1]: Entering directory '/home/ThibFrgsGmz/fprime/Ref/build-fprime-automatic-native/CMakeFiles/CMakeScratch/TryCompile-S9pOWR'
    Building CXX object CMakeFiles/cmTC_e755a.dir/testCXXCompiler.cxx.o
    /home/ThibFrgsGmz/ikos/build/install/bin/ikos-scan-c++    -MD -MT CMakeFiles/cmTC_e755a.dir/testCXXCompiler.cxx.o -MF CMakeFiles/cmTC_e755a.dir/testCXXCompiler.cxx.o.d -o CMakeFiles/cmTC_e755a.dir/testCXXCompiler.cxx.o -c /home/ThibFrgsGmz/fprime/Ref/build-fprime-automatic-native/CMakeFiles/CMakeScratch/TryCompile-S9pOWR/testCXXCompiler.cxx
    Linking CXX executable cmTC_e755a
    /usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_e755a.dir/link.txt --verbose=1
    /home/ThibFrgsGmz/ikos/build/install/bin/ikos-scan-c++ CMakeFiles/cmTC_e755a.dir/testCXXCompiler.cxx.o -o cmTC_e755a
    /usr/bin/ld: cannot find -lstdc++
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
    make[1]: *** [CMakeFiles/cmTC_e755a.dir/build.make:100: cmTC_e755a] Error 1
    make[1]: Leaving directory '/home/ThibFrgsGmz/fprime/Ref/build-fprime-automatic-native/CMakeFiles/CMakeScratch/TryCompile-S9pOWR'
    make: *** [Makefile:127: cmTC_e755a/fast] Error 2





  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:23 (project)


[ERROR] CMake erred with return code 1. Partial build cache remains. Run purge to clean-up.
```

**Comment by ThibFrgsGmz** [comment]
Well, if I bypass the compiler check by adding `set(CMAKE_TRY_COMPILE_TARGET_TYPE "STATIC_LIBRARY")` above `project()`, the `ikos-scan fprime-util generate` apparently works fine.

But when building, the compiler does not find the header files from `PlatformTypes.hpp` :

```bash
$ ikos-scan fprime-util build
[  0%] Building CXX object F-Prime/Fw/Cfg/CMakeFiles/Fw_Cfg.dir/ConfigCheck.cpp.o
In file included from /home/ThibFrgsGmz/fprime/Fw/Cfg/ConfigCheck.cpp:13:
In file included from /home/ThibFrgsGmz/fprime/config/FpConfig.hpp:14:
In file included from /home/ThibFrgsGmz/fprime/Fw/Types/BasicTypes.hpp:16:
/home/ThibFrgsGmz/fprime/cmake/platform/types/PlatformTypes.hpp:25:10: fatal error: 'limits' file not found
#include <limits>
         ^~~~~~~~
1 error generated.
make[3]: *** [F-Prime/Fw/Cfg/CMakeFiles/Fw_Cfg.dir/build.make:76: F-Prime/Fw/Cfg/CMakeFiles/Fw_Cfg.dir/ConfigCheck.cpp.o] Error 1
make[2]: *** [CMakeFiles/Makefile2:2539: F-Prime/Fw/Cfg/CMakeFiles/Fw_Cfg.dir/all] Error 2
make[1]: *** [CMakeFiles/Makefile2:1986: CMakeFiles/Ref.dir/rule] Error 2
make: *** [Makefile:195: Ref] Error 2
[ERROR] CMake erred with return code 2
```

In progress!

**Reply by LeStarch** [reply]
Does IKOS support C++11?  This header was added to support the C++11 limits setup.

**Reply by LeStarch** [reply]
Does IKOS support C++11?  This header was added to support the C++11 limits setup.

**Reply by ThibFrgsGmz** [reply]
Although I'm browsing the IKOS code base, I don't see a standard on the C++ being analyzed.
IKOS is done in C++14, and in one [file](https://github.com/NASA-SW-VnV/ikos/blob/a9789d20aa714c6698a37610f5c936406f48d379/frontend/llvm/test/regression/import/no_optimization/update-ll#L9) I see that they add `c++17` flag. 


Unless I'm mistaken, C++ is backwards compatible, but there may be a backwards compatibility problem at the toolchain level.

But in this case I don't think that the problem comes from the C++ version but that it comes upstream. Indeed if I comment the inclusion of `<limits>`, the compiler [does not find](https://github.com/nasa/fprime/blob/devel/cmake/platform/types/PlatformTypes.hpp#L37:L38) `<csting>`, nor `<cinttypes>`.


PS: did you see that there was a duplicate inclusion of `<limits>` in the file? Is it intentional?

**Reply by LeStarch** [reply]
If they are C++14/C++17 this should work.

Your problem is this line: https://github.com/nasa/fprime/blob/14091f63b9ff9d71ccc94eae65788b381a253b03/cmake/platform/Linux.cmake#L18

Here `PlatformTypes.hpp` is labeled as a `SYSTEM` include, which in some toolchains is treated as a C import not a C++ import.  🤦‍♀️ 

If you remove SYSTEM in that line it should work!

**Reply by ThibFrgsGmz** [reply]
Missed...😥I removed the `SYSTEM` keyword  (`include_directories("${CMAKE_CURRENT_LIST_DIR}/types")`) as you suggest but that doesn't let it know the location of `<limit>`.

```sh
/home/ThibFrgsGmz/fprime/cmake/platform/types/PlatformTypes.hpp:25:10: fatal error: 'limits' file not found
```

I'll have to dive into these CMake configuration files.

**Reply by LeStarch** [reply]
If the above did not fix it then it is a problem with the tool's C++ setup or the file is somehow being included in a C file. I'd rule-out the latter as we'd have see other issues, but you could check based on the "included from" info in the compile output.

**Reply by ThibFrgsGmz** [reply]
Well I will study MISRA with cppcheck and/or Cobra to change minds. 😅

**Comment by ThibFrgsGmz** [comment]
@LeStarch FYI, the problem on the impossibility to find the system headers would come from my version of `gcc`.

I tried to run `clang-tidy-12` on the F' codebase and I got the same error as when I ran IKOS.

I then changed my version of `gcc` from `9.x` to `10.x` and the problem of `clang-tidy-12` does not appear any more.

However, in the meantime I had deleted my IKOS installation for the last one supporting LLVM 14 and I can't even get the build or download via Homebrew... I have opened a ticket with them: https://github.com/NASA-SW-VnV/homebrew-core/issues/5
